File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/UserService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.dto.ChangeUserPasswordDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.LoginDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.PageableDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.SignUpDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.UserDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.UserInfoDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.ApplicationUser;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;

public interface UserService extends UserDetailsService {

    /**
     * Find a user in the context of Spring Security based on the email address
     * <br>
     * For more information have a look at this tutorial:
     * https://www.baeldung.com/spring-security-authentication-with-a-database
     *
     * @param email the email address
     * @return a Spring Security user
     * @throws UsernameNotFoundException is thrown if the specified user does not exists
     */
    @Override
    UserDetails loadUserByUsername(String email) throws UsernameNotFoundException;

    /**
     * Find an application user based on the email address.
     *
     * @param email the email address
     * @return a application user
     */
    ApplicationUser findApplicationUserByEmail(String email);

    /**
     * Find an application user based on the username.
     *
     * @param username the email address
     * @return a application user
     */
    ApplicationUser findApplicationUserByUsername(String username);

    /**
     * Find an application user based on the id.
     *
     * @param id the id
     * @return an application user
     */
    ApplicationUser findApplicationUserById(Long id);

    /**
     * Log in a user.
     *
     * @param loginDto login credentials
     * @return the JWT, if successful
     * @throws org.springframework.security.authentication.BadCredentialsException if credentials are bad
     */
    String login(LoginDto loginDto);

    /**
     * Generates a new token for the given user.
     *
     * @param email email of the given user
     * @return the JWT, if successful
     * @throws BadCredentialsException if credentials are bad
     */
    String regenerateToken(String email);

    /**
     * Lock a user.
     *
     * @param user attempted user
     * @return user locked after increase
     * @throws org.springframework.security.authentication.BadCredentialsException if credentials are bad
     */
    boolean increaseLoginAttempts(ApplicationUser user);

    /**
     * Sign Up a user.
     *
     * @param signUpDto signUp credentials
     * @return the JWT
     */
    String signUp(SignUpDto signUpDto);

    /**
     * Returns the user id for the given email.
     *
     * @param email email of the location to return
     * @return UserDto Location with the given id
     */
    Long getUserIdByEmail(String email);

    /**
     * Returns the user with the given id.
     *
     * @param id ID of the user to return
     * @return UserDto User with the given id
     */
    UserDto getUserById(Long id);

    /**
     * Updates the User with the given id (only the user and admin can update the User information).
     *
     * @param id      ID of the ApplicationUser to update
     * @param userDto User to update
     * @return UserDto user with the given id
     */
    UserDto updateUser(Long id, UserDto userDto);

    /**
     * Changes the User Password to a new Password.
     *
     * @param id  ID of the ApplicationUser to update
     * @param dto Password Confirmation and New Passwords
     */
    void changeUserPassword(Long id, ChangeUserPasswordDto dto);

    /**
     * Returns all users in a pageable format.
     *
     * @param locked   whether to return only locked users
     * @param page     the page number
     * @param pageSize the size of the page
     * @return UserDto[] all users
     */
    PageableDto<UserDto> getAllUsers(boolean locked, int page, int pageSize);

    /**
     * Locks the given user.
     *
     * @param id the ID of the user to lock
     */
    void lockUser(Long id);

    /**
     * Unlocks the given user.
     *
     * @param id the ID of the user to unlock
     */
    void unlockUser(Long id);

    /**
     * Returns the email of the admin.
     *
     * @return the email of the admin
     */
    String getAdminEmail();

    /**
     * Returns short info of calling user.
     *
     * @return info of calling user
     */
    UserInfoDto getUserInfoOfCallingUser();
}

File: frontend/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { Globals } from '../global';
import { Pageable, UserDto, UserInfo, UserPasswordChangeDto } from '../dtos';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private userBaseUri: string = this.globals.backendUri + '/user';
  private changePasswordUri = 'change-password';

  constructor(private httpClient: HttpClient, private globals: Globals) {}

  /**
   * This searches for a user given his id
   *
   * @param id The id of the user
   * @returns The user if found
   */
  getOneById(id: number): Observable<UserDto> {
    console.log(`Get user with id=${id}`);
    const url = `${this.userBaseUri}/${id}`;
    return this.httpClient.get<UserDto>(url);
  }

  /**
   * This searches for a user given by his email
   *
   * @param id The id of the user
   * @returns The user if found
   */
  getOneByEmail(email: string): Observable<number> {
    console.log(`Get user with email=${email}`);
    return this.httpClient.get<number>(`${this.userBaseUri}`, {
      params: { email },
    });
  }

  /**
   * Get short info of logged-in user
   *
   * @returns short info of the user
   */
  getInfoOfLoggedInUser(): Observable<UserInfo> {
    return this.httpClient.get<UserInfo>(`${this.userBaseUri}/info`);
  }

  /**
   * This updates the password for a user given by valid password and new valid password
   *
   * @param id The id of the user
   * @param dto The passwords
   */
  changeUserPassword(id: number, dto: UserPasswordChangeDto) {
    console.log(`Update password for user with id = ${id}`);
    return this.httpClient.put(`${this.userBaseUri}/${id}/${this.changePasswordUri}`, dto);
  }

  /**
   * Updates the user with the given id with the given data.
   *
   * @param id The id of the user to update
   * @param user The user data to update
   * @returns The updated user
   */
  update(id: number, user: UserDto): Observable<UserDto> {
    return this.httpClient.put<UserDto>(`${this.userBaseUri}/${id}`, user);
  }

  /**
   * Deletes a given user.
   *
   * @param id ID of the given user
   * @returns observable of void, only useful for error handling
   */
  deleteUser(id: number): Observable<void> {
    return this.httpClient.delete<void>(this.userBaseUri + '/' + id);
  }

  /**
   * Returns all users (locked and unlocked)
   *
   * @param locked true if only locked users should be returned
   * @param page the page to return
   * @param pageSize the size of the page
   * @returns all users
   */
  getAllUsers(locked: boolean, page: number, pageSize: number): Observable<Pageable<UserDto>> {
    return this.httpClient.get<Pageable<UserDto>>(this.userBaseUri + '/all', { params: { locked, page, pageSize } });
  }

  /**
   * Locks a user given by his ID (no admin is lockable)
   *
   * @param id The id of the user to lock
   * @returns observable of void, only useful for error handling
   */
  lockUser(id: number): Observable<void> {
    return this.httpClient.put<void>(this.userBaseUri + '/lock/' + id, null);
  }

  /**
   * Unlocks a user given by his ID
   *
   * @param id The id of the user to unlock
   * @returns observable of void, only useful for error handling
   */
  unlockUser(id: number): Observable<void> {
    return this.httpClient.put<void>(this.userBaseUri + '/unlock/' + id, null);
  }

  /**
   * Returns an email address of an admin
   *
   * @returns email address of an admin
   */
  getAdminEmail(): Observable<string> {
    return this.httpClient.get<string>(this.globals.backendUri + '/authentication/admin-email', {
      responseType: 'text' as 'json',
    });
  }
}
